### PR TITLE
Try to fix DPCPU/VNET relocations in purecap kernels

### PR DESCRIPTION
**Pull request by markjdb**
This fixes a problem I was seeing where vmm.ko's use of a DPCPU pointer was yielding exceptions due to `_DPCPU_PTR()` giving an invalid capability. It turned out that the R_MORELLO_RELATIVE relocation handler did not apply the transformation in elf_relocaddr() needed for `_DPCPU_PTR` to work. This patch tries to fix the problem.

I'm sure my handling of the addend here doesn't make sense, but I'm not sure what the right behaviour is. There is no addend in the particular relocation that was causing me problems. Probably I should pass `addr + addend` to elf_relocaddr() and continue to use `addr` as the base address when dealing with a DPCPU/VNET symbol?

### CODE REVIEW

File: sys/arm64/arm64/elf_machdep.c [hunk 1]
```diff
@@ -476,11 +488,32 @@ elf_reloc_internal(linker_file_t lf, char *relocbase, const void *data,
...
+			addr1 = elf_relocaddr(lf, addr + addend);
+			if (addr + addend == addr1) {
+				addr1 = addr;
+				base = (__cheri_tocap void * __capability)
+				    relocbase;
+			} else {
+				addend = 0;
+				base = (__cheri_tocap void * __capability)
+				    linker_kernel_file->address;
+			}
```
Comment (jrtc27) [inline sys/arm64/arm64/elf_machdep.c:514]: Given what elf_relocaddr does, I think the simplest thing to do that works is just this?

```suggestion
			addr = elf_relocaddr(lf, addr + addend) - addend;
```

Though arguably if that's not equivalent to the following then the input was very dodgy:

```suggestion
			addr = elf_relocaddr(lf, addr);
```

Unlike normal R_FOO_RELATIVE, where the addend is the offset from the entire start of the DSO's mapping, for R_MORELLO_RELATIVE it's the offset from the capability's base address, which really should reside within the regions if and only if the address as a whole does.

Comment (markjdb) [reply sys/arm64/arm64/elf_machdep.c:514]: > for R_MORELLO_RELATIVE it's the offset from the capability's base address, which really should reside within the regions if and only if the address as a whole does.

The last part wasn't really obvious to me from the Morello ABI docs, but it'd be pretty surprising if it weren't true.

File: sys/arm64/arm64/elf_machdep.c [hunk 2]
```diff
@@ -377,22 +377,24 @@ reloc_instr_imm(Elf32_Addr *where, Elf_Addr val, u_int msb, u_int lsb)
 }
 
 #if __has_feature(capabilities)
+static void __nosanitizecoverage
+decode_fragment(Elf_Addr *fragment, Elf_Addr relocbase, Elf_Addr *addrp,
+    Elf_Addr *sizep, uint8_t *permsp)
+{
+	*addrp = fragment[0] + relocbase;
```
Comment (bsdjhb) [inline sys/arm64/arm64/elf_machdep.c:384]: Normally we prefer to put offsets on the RHS.

**Review by bsdjhb — APPROVED** [review]

**Review by bsdjhb — APPROVED** [review]
Is there a simpler test case we can use to fix this on RISC-V as well?  Or maybe just open an issue against CheriBSD for fixing this with RISC-V with a description of a simple enough test case (probably just a DPCPU value that's a pointer?)

**Comment by markjdb** [comment]
> Is there a simpler test case we can use to fix this on RISC-V as well? Or maybe just open an issue against CheriBSD for fixing this with RISC-V with a description of a simple enough test case (probably just a DPCPU value that's a pointer?)

Since this affects VNETs as well there is a very simple test case: `ifconfig epair create`. Without this patch, the kernel crashes on my Morello box with:

```
panic: Capability abort from kernel space: tag violation
cpuid = 3
time = 1698772342
KDB: stack backtrace:
db_trace_self() at db_trace_self
db_trace_self_wrapper() at db_trace_self_wrapper+0x44
vpanic() at vpanic+0x1e8
panic() at panic+0x48
cap_abort() at cap_abort+0x248
handle_el1h_sync() at handle_el1h_sync+0x10
--- exception, esr 0x96000068
vnet_epair_init() at vnet_epair_init+0x78
linker_load_module() at linker_load_module+0xca4
kern_kldload() at kern_kldload+0x1c0
sys_kldload() at sys_kldload+0xa0
do_el0_sync() at do_el0_sync+0x72c
handle_el0_sync() at handle_el0_sync+0x2c
--- exception, esr 0x56000000
```

Indeed, `vnet_epair_init()` stores to a per-VNET pointer.

I'm quite short on time, so will just file an issue for now since I don't have an environment ready to go for testing riscv.

**Pull request merged**